File: app/src/main/java/com/example/ui/alarm/AlarmService.java
package com.example.ui.alarm;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.media.AudioAttributes;
import android.media.MediaPlayer;
import android.media.RingtoneManager;
import android.net.Uri;
import android.os.Build;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.os.VibrationEffect;
import android.os.Vibrator;
import android.os.VibratorManager;
import android.util.Log;
import androidx.core.app.NotificationCompat;
import com.example.MainActivity;
import com.example.data.database.AppDatabase;
import com.example.data.model.Alarm;
import java.util.concurrent.Executors;

public class AlarmService extends Service {
    private static final String TAG = "AlarmService";
    private static final int NOTIFICATION_ID = 9999;
    private static final String CHANNEL_ID = "OWNUP_ALARM_CHANNEL";

    private MediaPlayer mediaPlayer = null;
    private Vibrator vibrator = null;
    private Alarm currentAlarm = null;
    private final Handler handler = new Handler(Looper.getMainLooper());
    private Runnable autoStopRunnable = null;

    @Override
    public void onCreate() {
        super.onCreate();
        Log.d(TAG, "AlarmService created");
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        int alarmId = intent != null ? intent.getIntExtra("ALARM_ID", -1) : -1;
        Log.d(TAG, "onStartCommand with Alarm ID: " + alarmId);

        if (alarmId != -1) {
            startForegroundServiceFlow(alarmId);
        } else {
            stopSelf();
        }

        return START_NOT_STICKY;
    }

    private void startForegroundServiceFlow(int alarmId) {
        // First, build a quick placeholder notification so startForeground doesn't fail on newer Android versions
        createNotificationChannel();
        Notification notification = buildNotification("Morning Wake Up Alarm", "Time to own your morning!", null);
        startForeground(NOTIFICATION_ID, notification);

        // Fetch alarm on a background thread and then post back to the main thread
        Executors.newSingleThreadExecutor().execute(() -> {
            AppDatabase db = AppDatabase.getDatabase(getApplicationContext());
            Alarm alarm = db.alarmDao().getAlarmById(alarmId);

            handler.post(() -> {
                if (alarm != null) {
                    currentAlarm = alarm;
                    Log.d(TAG, "Found alarm: " + alarm.getTaskType());

                    // Update global alarm state holder
                    AlarmStateHolder.startRinging(alarm);

                    // Update notification with real details
                    Notification updatedNotification = buildNotification(
                        "ALARM RINGING: " + alarm.getTaskType(),
                        alarm.getTaskDetails(),
                        alarm
                    );
                    NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
                    if (notificationManager != null) {
                        notificationManager.notify(NOTIFICATION_ID, updatedNotification);
                    }

                    // Play Audio and Vibrate
                    startAudio(alarm);
                    startVibration(alarm);

                    // Start automatic ring duration timeout
                    scheduleAutoStop(alarm);
                } else {
                    Log.e(TAG, "Alarm ID " + alarmId + " not found in database");
                    stopSelf();
                }
            });
        });
    }

    private void scheduleAutoStop(final Alarm alarm) {
        cancelAutoStop();

        long durationMs = alarm.getRingDurationMinutes() * 60 * 1000L;
        autoStopRunnable = () -> {
            Log.d(TAG, "Auto-stop duration reached. Moving to Focus Mode!");
            
            // Auto transition to Focus Mode even if user missed it
            AlarmStateHolder.startFocusMode(
                alarm.getId(),
                alarm.getActivityDurationMinutes(),
                alarm.getTaskType(),
                alarm.getTaskDetails(),
                alarm.getReferencePhotoPath()
            );
            AlarmStateHolder.stopRinging();
            stopSelf();
        };

        handler.postDelayed(autoStopRunnable, durationMs);
    }

    private void cancelAutoStop() {
        if (autoStopRunnable != null) {
            handler.removeCallbacks(autoStopRunnable);
            autoStopRunnable = null;
        }
    }

    private void startAudio(Alarm alarm) {
        try {
            Uri ringtoneUri = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_ALARM);
            if (ringtoneUri == null) {
                ringtoneUri = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_RINGTONE);
            }

            if (mediaPlayer != null) {
                mediaPlayer.stop();
                mediaPlayer.release();
            }

            mediaPlayer = new MediaPlayer();
            mediaPlayer.setDataSource(getApplicationContext(), ringtoneUri);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                mediaPlayer.setAudioAttributes(
                    new AudioAttributes.Builder()
                        .setUsage(AudioAttributes.USAGE_ALARM)
                        .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                        .build()
                );
            } else {
                mediaPlayer.setAudioStreamType(android.media.AudioManager.STREAM_ALARM);
            }
            mediaPlayer.setLooping(true);
            mediaPlayer.prepare();
            mediaPlayer.start();
        } catch (Exception e) {
            Log.e(TAG, "Failed to play audio", e);
        }
    }

    private void startVibration(Alarm alarm) {
        if (!alarm.isVibrate()) return;

        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                VibratorManager vibratorManager = (VibratorManager) getSystemService(Context.VIBRATOR_MANAGER_SERVICE);
                vibrator = vibratorManager.getDefaultVibrator();
            } else {
                vibrator = (Vibrator) getSystemService(Context.VIBRATOR_SERVICE);
            }

            if (vibrator == null) return;

            long[] pattern = {0, 800, 400, 800};
            int[] amplitudes = {0, 255, 0, 255};

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                vibrator.vibrate(VibrationEffect.createWaveform(pattern, amplitudes, 0));
            } else {
                vibrator.vibrate(pattern, 0);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to start vibration", e);
        }
    }

    private Notification buildNotification(String title, String text, Alarm alarm) {
        Intent fullScreenIntent = new Intent(getApplicationContext(), MainActivity.class);
        fullScreenIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
        fullScreenIntent.putExtra("LAUNCH_RINGING_ALARM", true);
        if (alarm != null) {
            fullScreenIntent.putExtra("ALARM_ID", alarm.getId());
        }

        int pendingIntentFlags = PendingIntent.FLAG_UPDATE_CURRENT;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            pendingIntentFlags |= PendingIntent.FLAG_IMMUTABLE;
        }

        PendingIntent fullScreenPendingIntent = PendingIntent.getActivity(
            getApplicationContext(),
            0,
            fullScreenIntent,
            pendingIntentFlags
        );

        return new NotificationCompat.Builder(getApplicationContext(), CHANNEL_ID)
            .setSmallIcon(android.R.drawable.ic_lock_idle_alarm)
            .setContentTitle(title)
            .setContentText(text)
            .setPriority(NotificationCompat.PRIORITY_MAX)
            .setCategory(NotificationCompat.CATEGORY_ALARM)
            .setFullScreenIntent(fullScreenPendingIntent, true)
            .setContentIntent(fullScreenPendingIntent)
            .setOngoing(true)
            .setAutoCancel(false)
            .build();
    }

    private void createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(
                CHANNEL_ID,
                "OwnUp Alarm Alerts",
                NotificationManager.IMPORTANCE_HIGH
            );
            channel.setDescription("Channels for ringing wake-up alarms in OwnUp");
            channel.enableLights(true);
            channel.enableVibration(true);
            channel.setBypassDnd(true); // Crucial to ring in Silent/DND mode!

            NotificationManager manager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
            if (manager != null) {
                manager.createNotificationChannel(channel);
            }
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        Log.d(TAG, "AlarmService destroyed");
        cancelAutoStop();

        try {
            if (mediaPlayer != null) {
                mediaPlayer.stop();
                mediaPlayer.release();
                mediaPlayer = null;
            }
        } catch (Exception e) {
            Log.e(TAG, "Error cleaning up MediaPlayer", e);
        }

        try {
            if (vibrator != null) {
                vibrator.cancel();
                vibrator = null;
            }
        } catch (Exception e) {
            Log.e(TAG, "Error cleaning up Vibrator", e);
        }
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }
}

File: app/src/main/java/com/example/ui/alarm/AlarmService.kt
package com.example.ui.alarm

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.media.AudioAttributes
import android.media.MediaPlayer
import android.media.RingtoneManager
import android.os.Build
import android.os.IBinder
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import android.util.Log
import androidx.core.app.NotificationCompat
import com.example.MainActivity
import com.example.data.database.AppDatabase
import com.example.data.model.Alarm
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class AlarmService : Service() {
    private val serviceJob = Job()
    private val serviceScope = CoroutineScope(Dispatchers.Main + serviceJob)

    private var mediaPlayer: MediaPlayer? = null
    private var vibrator: Vibrator? = null
    private var currentAlarm: Alarm? = null
    private var autoStopJob: Job? = null

    private val NOTIFICATION_ID = 9999
    private val CHANNEL_ID = "OWNUP_ALARM_CHANNEL"

    override fun onCreate() {
        super.onCreate()
        Log.d("AlarmService", "AlarmService created")
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        val alarmId = intent?.getIntExtra("ALARM_ID", -1) ?: -1
        Log.d("AlarmService", "onStartCommand with Alarm ID: $alarmId")

        if (alarmId != -1) {
            startForegroundServiceFlow(alarmId)
        } else {
            stopSelf()
        }

        return START_NOT_STICKY
    }

    private fun startForegroundServiceFlow(alarmId: Int) {
        // First, build a quick placeholder notification so startForeground doesn't fail on newer Android versions
        createNotificationChannel()
        val notification = buildNotification("Morning Wake Up Alarm", "Time to own your morning!", null)
        startForeground(NOTIFICATION_ID, notification)

        // Then, fetch the real alarm in a coroutine and customize
        serviceScope.launch {
            val db = AppDatabase.getDatabase(applicationContext)
            val alarm = withContext(Dispatchers.IO) {
                db.alarmDao().getAlarmById(alarmId)
            }

            if (alarm != null) {
                currentAlarm = alarm
                Log.d("AlarmService", "Found alarm: ${alarm.taskType}")

                // Update global alarm state holder
                AlarmStateHolder.startRinging(alarm)

                // Update notification with real details
                val updatedNotification = buildNotification(
                    title = "ALARM RINGING: ${alarm.taskType}",
                    text = alarm.taskDetails,
                    alarm = alarm
                )
                val notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
                notificationManager.notify(NOTIFICATION_ID, updatedNotification)

                // Play Audio and Vibrate
                startAudio(alarm)
                startVibration(alarm)

                // Start automatic ring duration timeout
                scheduleAutoStop(alarm)
            } else {
                Log.e("AlarmService", "Alarm ID $alarmId not found in database")
                stopSelf()
            }
        }
    }

    private fun scheduleAutoStop(alarm: Alarm) {
        autoStopJob?.cancel()
        autoStopJob = serviceScope.launch {
            // Wait for ring duration (default 5 minutes)
            val durationMs = alarm.ringDurationMinutes * 60 * 1000L
            delay(durationMs)
            Log.d("AlarmService", "Auto-stop duration reached. Moving to Focus Mode!")
            
            // Auto transition to Focus Mode even if user missed it
            AlarmStateHolder.startFocusMode(
                alarmId = alarm.id,
                durationMinutes = alarm.activityDurationMinutes,
                taskType = alarm.taskType,
                taskDetails = alarm.taskDetails
            )
            AlarmStateHolder.stopRinging()
            stopSelf()
        }
    }

    private fun startAudio(alarm: Alarm) {
        try {
            val ringtoneUri = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_ALARM)
                ?: RingtoneManager.getDefaultUri(RingtoneManager.TYPE_RINGTONE)

            mediaPlayer?.stop()
            mediaPlayer?.release()

            mediaPlayer = MediaPlayer().apply {
                setDataSource(applicationContext, ringtoneUri)
                setAudioAttributes(
                    AudioAttributes.Builder()
                        .setUsage(AudioAttributes.USAGE_ALARM)
                        .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                        .build()
                )
                isLooping = true
                prepare()
                start()
            }
        } catch (e: Exception) {
            Log.e("AlarmService", "Failed to play audio", e)
        }
    }

    private fun startVibration(alarm: Alarm) {
        if (!alarm.isVibrate) return

        try {
            vibrator = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                val vibratorManager = getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager
                vibratorManager.defaultVibrator
            } else {
                @Suppress("DEPRECATION")
                getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
            }

            val pattern = longArrayOf(0, 800, 400, 800)
            val amplitudes = intArrayOf(0, 255, 0, 255)

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                vibrator?.vibrate(VibrationEffect.createWaveform(pattern, amplitudes, 0))
            } else {
                @Suppress("DEPRECATION")
                vibrator?.vibrate(pattern, 0)
            }
        } catch (e: Exception) {
            Log.e("AlarmService", "Failed to start vibration", e)
        }
    }

    private fun buildNotification(title: String, text: String, alarm: Alarm?): Notification {
        val fullScreenIntent = Intent(applicationContext, MainActivity::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
            putExtra("LAUNCH_RINGING_ALARM", true)
            if (alarm != null) {
                putExtra("ALARM_ID", alarm.id)
            }
        }

        val fullScreenPendingIntent = PendingIntent.getActivity(
            applicationContext,
            0,
            fullScreenIntent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )

        return NotificationCompat.Builder(applicationContext, CHANNEL_ID)
            .setSmallIcon(android.R.drawable.ic_lock_idle_alarm)
            .setContentTitle(title)
            .setContentText(text)
            .setPriority(NotificationCompat.PRIORITY_MAX)
            .setCategory(NotificationCompat.CATEGORY_ALARM)
            .setFullScreenIntent(fullScreenPendingIntent, true)
            .setContentIntent(fullScreenPendingIntent)
            .setOngoing(true)
            .setAutoCancel(false)
            .build()
    }

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                CHANNEL_ID,
                "OwnUp Alarm Alerts",
                NotificationManager.IMPORTANCE_HIGH
            ).apply {
                description = "Channels for ringing wake-up alarms in OwnUp"
                enableLights(true)
                enableVibration(true)
                setBypassDnd(true) // Crucial to ring in Silent/DND mode!
            }

            val manager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            manager.createNotificationChannel(channel)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.d("AlarmService", "AlarmService destroyed")
        autoStopJob?.cancel()
        serviceJob.cancel()

        try {
            mediaPlayer?.stop()
            mediaPlayer?.release()
            mediaPlayer = null
        } catch (e: Exception) {
            Log.e("AlarmService", "Error cleaning up MediaPlayer", e)
        }

        try {
            vibrator?.cancel()
            vibrator = null
        } catch (e: Exception) {
            Log.e("AlarmService", "Error cleaning up Vibrator", e)
        }
    }

    override fun onBind(intent: Intent?): IBinder? {
        return null
    }
}
